### PR TITLE
Framework: Better babel config for node

### PR DESCRIPTION
**Pull request by blowery**
Transpile fewer things with the env preset

### CODE REVIEW

**Comment by jsnajdr** [comment]
Investigating the e2e failures, I discovered that setting `loose: true` for all Babel transforms is not a good idea. When transpiling array spreads in loose mode, Babel assumes that the elements being spread are arrays. That breaks code where we spread iterators:
```js
let m = new Map();
let a = [ ...m.values() ];
```
This code transpiles to something that fails on a non-array (iterator doesn't have `concat` method):
```js
var a = m.values().concat();
```
This really happens in Calypso when [showing domain suggestions](https://github.com/Automattic/wp-calypso/blob/d544e149e535c20072956bb6bcac637163a0e807/client/components/domains/register-domain-step/index.jsx#L740) during signup, breaking e2e tests.

**Comment by jsnajdr** [comment]
I tried to remove the `loose` options for all transforms, including the ones that previously had it (class properties, template literals). Are the bundle size savings worth the trouble of having to know the subtle differences and worrying about whether their break our app or not? Especially with the Calypso SDK, I'd rather be on the safe side of things.

To the `build` chunk, this change adds:
- parsed: +14849 B (+1.1%)
- gzipped: +1783 B (+0.6%)

To the complete load of `build` + `reader` chunk groups, this change adds:
- parsed: +25590 B (+0.9%)
- gzipped: +2854 B (+0.4%)

http://iscalypsofastyet.com/push/83280fdcd2a5b9e17df296d5be45bd7b0f3dda3e
http://iscalypsofastyet.com/chunkgroups?branch=update/better-babel-config-for-server

**Comment by sirreal** [comment]
> I'd rather be on the safe side of things.

Seeing how easily and silently this broke, I'm inclined to be on the safe side as well.

**Comment by sirreal** [comment]
Full e2e is failing regularly:

https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/13214

Can you investigate @jsnajdr?

**Comment by jsnajdr** [comment]
> Full e2e is failing regularly:

These failures are related to editor and uploading images -- and indeed, trying to update feature image locally leads to infinite loop. Debugging.

**Comment by blowery** [comment]
Tests passed woo

**Review by blowery — COMMENTED** [review]
I'd approve this, but as the original author I can't, technically. But ship when ready!

File: babel.config.js [hunk 1]
```diff
@@ -7,29 +7,24 @@ const isCalypsoClient = process.env.CALYPSO_CLIENT === 'true';
 const modules = isCalypsoClient ? false : 'commonjs'; // only calypso should keep es6 modules
 const codeSplit = require( './server/config' ).isEnabled( 'code-splitting' );
 
+const targets = isCalypsoClient
+	? { browsers: [ 'last 2 versions', 'Safari >= 10', 'iOS >= 10', 'ie >= 11' ] }
+	: { node: 'current' };
```
Comment (sirreal) [inline babel.config.js]: This is the version of node where the process is running. Perfect 👍 

As a reader, I had interpreted this as something like "latest stable." That is _not_ the case:

> #### `targets.node`
> 
> `string | "current" | true`.
> 
> If you want to compile against the current node version, you can specify `"node": true` or `"node": "current"`, which would be the same as `"node": process.versions.node`.

https://babeljs.io/docs/en/babel-preset-env#targetsnode

**Review by sirreal — APPROVED** [review]
I'm happy to see this simplification and with full e2e passing let's 🚢 

Thanks!

**Comment by jsnajdr** [comment]
Thanks for reviewing @blowery and @sirreal! When working with this patch applied locally, I see the Node process that runs Calypso in development mode (after `npm start`) crashing very often with out-of-memory exception. The V8 error message points somewhere into `Stats.js` writer in webpack.

I'll investigate this issue a little before merging this PR. I don't want to break development mode for everyone :wink:

**Comment by jsnajdr** [comment]
I failed to find anything when searching for the cause of the sudden memory consumption increase. The default max heap size in Node is around 1.5GB and that's not much given we're doing a full webpack build inside. Let's just increase the max heap size to 8GB. I tried to avoid this.

**Comment by blowery** [comment]
Ditto, I could repro this dying constantly while in dev. I noticed that we're not setting the max heap size for the dev run, so ...

```
diff --git a/package.json b/package.json
index fba660ce77..d160ca3f7d 100644
--- a/package.json
+++ b/package.json
@@ -252,7 +252,7 @@
     "rm": "node bin/rm.js",
     "prestart": "npx check-node-version --package && npm run -s install-if-deps-outdated && node bin/welcome.js",
     "start": "npm run -s build",
-    "poststart": "cross-env-shell CALYPSO_CLIENT=true NODE_PATH=$NODE_PATH:server:client:. node $NODE_ARGS build/bundle.js",
+    "poststart": "cross-env-shell CALYPSO_CLIENT=true NODE_PATH=$NODE_PATH:server:client:. node --max_old_space_size=8192 $NODE_ARGS build/bundle.js",
     "reformat-files": "./node_modules/.bin/prettier --ignore-path .eslintignore --write --require-pragma \"**/*.{js,jsx,scss}\"",
     "pretest": "npm run -s install-if-deps-outdated",
     "test": "run-s -s test-client test-server",
```

seems to help.

**Comment by blowery** [comment]
@jsnajdr oof, that's breaking the build in circle... likely eating too much memory for the container. :/

**Comment by blowery** [comment]
@sirreal you ran into this prior i think?

**Comment by blowery** [comment]
trying 4G instead of 8. Seems to be ok on rebuilds locally.

**Comment by jsnajdr** [comment]
I experienced several crashes in the last few days even on other branches that don't have this Babel patch. The increased crash frequency is likely caused by something else.

I also verified that the server bundle size (`build/bundle.js`) didn't get bigger, but rather significantly smaller: 5.6 MB to 4.8 MB.

Going to merge now.

**Pull request merged**